default/python/AI/AIFleetMission.py

```diff
@@ -268,7 +268,19 @@ def _check_abort_mission(self, fleet_order: AIFleetOrder):
             elif isinstance(fleet_order, OrderOutpost):
                 if planet.unowned:
                     return False
-            elif isinstance(fleet_order, OrderInvade):  # TODO add substantive abort check
+            elif isinstance(fleet_order, OrderInvade):
+                if planet.owner == fo.getEmpire().empireID:
+                    debug(f"Abandoning invasion for {planet}, it already belongs to us")
+                    return True
+                if planet.currentMeterValue(fo.meterType.maxShield):
+                    military_support_fleets = MilitaryAI.get_military_fleets_with_target_system(planet.systemID)
+                    if not military_support_fleets:
+                        # Maybe try again later...
+                        debug(
+                            f"Abandoning Invasion mission for {planet} because target has nonzero max shields "
+                            f"and there is no military fleet assigned to secure the target system."
+                        )
```

Comment:
Please do not make big and complicated functions bigger and more complicated.

Just extract this code to a method. `_check_abort_mission_for_invarions_fleet`.

PS.  good OOP pattern is to replace these is instance calls with a single call of the fleet_order.should_abort() and move code from branches to classes.  But this is out of scope of this PR.

Reply:
Looking a fleet orders I wonder whether we could simply use fleet_order.is_valid. There is definitely some code duplication here.
But that is not a trivial refactoring, I suggest merging this as it is. _check_abort_mission is not that bad yet.